tests/cpp/test_multidevice_overlap.cpp

```diff
@@ -0,0 +1,242 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ATen/Functions.h>
+#include <c10/cuda/CUDAStream.h>
+#include <c10/util/ArrayRef.h>
+#include <fusion.h>
+#include <ir/utils.h>
+#include <ops/all_ops.h>
+#include <tests/cpp/multidevice.h>
+
+namespace nvfuser {
+
+struct OverlapTestParams {
+  // Tensors sizes
+  int64_t M = std::pow(2, 6);
+  int64_t K = std::pow(2, 5);
+  int64_t N = std::pow(2, 4);
+  int64_t S = std::pow(2, 3);
+
+  // network backend type
+  CommunicatorBackend backend_type = CommunicatorBackend::nccl;
+
+  // overlap optimization parameters
+  bool use_different_streams =
+      true; // whether to change CUDA stream at each iteration
+};
+
+std::ostream& operator<<(std::ostream& out, const OverlapTestParams& params) {
+  std::string indent = "  ";
+  out << "params:{\n"
+      << indent << "backend_type=" << params.backend_type << "\n"
+      << indent << "M=" << params.M << "\n"
+      << indent << "K=" << params.K << "\n"
+      << indent << "N=" << params.N << "\n"
+      << indent << "S=" << params.S << "\n"
+      << indent << "use_different_streams=" << params.use_different_streams
+      << "\n"
+      << "}";
+  return out;
+}
+
+class OverlapTest : public MultiDeviceTest {
+ protected:
+  OverlapTestParams params;
+
+  int64_t num_devices_;
+  int64_t my_device_index_;
+  at::Tensor ta_, tb_, tc_locally_reduced_, tc_, tc_expected_;
+  // stores the backend
+  c10d::Backend* world_communicator_;
+
+  void SetUp() {
+    MultiDeviceTest::SetUp();
+
+    num_devices_ = communicator->size();
+    my_device_index_ = communicator->deviceId();
+    ASSERT_EQ(params.M % (params.S * num_devices_), 0);
+    ASSERT_EQ(params.K % num_devices_, 0);
+
+    // Setup the world communicators
+    std::vector<int64_t> devices(num_devices_);
+    std::iota(devices.begin(), devices.end(), 0);
+    world_communicator_ =
+        communicator->getBackendForTeam(devices, params.backend_type);
+
+    // Define I/O and intermediate Tensor shapes
+    // clang-format off
+        std::vector<int64_t> ta_unsharded_sizes       = {params.S, num_devices_, params.M/params.S, params.K/num_devices_, 1       };
```

Comment:
```suggestion
    std::vector<int64_t> ta_unsharded_sizes       = {params.S, num_devices_, params.M/params.S, params.K/num_devices_, 1       };
```

Reply:
Nit: I don't see much value for `clang-format off`. You probably wanted to align commas, but personally I don't find it as annoying as having to indent manually :)

Reply:
ok, let me remove it then